### PR TITLE
[BREAKING] Use proper SAR for signed right shifts and emulate on pre-constantinople.

### LINKED ISSUE #3847: Signed rights shifts use SDIV and as such round differently to proper SAR
From https://github.com/ethereum/solidity/pull/3797/files#r179776071.

This could be changed in a breaking release.

### PR DESCRIPTION
**Pull request by ekpyron**
Closes #3847.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ### 0.5.0 (unreleased)
 
 Features:
+ * General: Signed right shift uses proper arithmetic shift, i.e. rounding towards negative infinity.
```
Comment (chriseth) [inline Changelog.md]: For 0.4.0 we had a separate section with breaking changes, I think we should do that again.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -1718,11 +1718,31 @@ void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator, Type co
...
+			m_context << Instruction::SAR;
 		else
-			m_context << u256(2) << Instruction::EXP << Instruction::SWAP1 << (c_valueSigned ? Instruction::SDIV : Instruction::DIV);
+		{
+			// if signed and negative, calculate
+			// ~((~value_to_shift) / (2**shift_amount))
+			if (c_valueSigned)
+			{
+				// stack: value_to_shift shift_amount
+				m_context << u256(0) << Instruction::DUP3 << Instruction::SLT;
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Can you use inline assembly for this?

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Actually, this can be done as a pragma 050, should we do that?

**Review by chriseth — DISMISSED** [review]
Please change it to using inline assembly.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 2 above)
Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Ok - I didn't use inline assembly so far, because this way it's clearer to me that the amount of overhead introduced is not too excessive. But I can change it now.

Regarding a pragma 050, I'm not sure - this change does not only change some warnings to errors and it isn't merely a breaking syntactic change, but it will actually semantically change existing code. Did we do that before with a pragma 050? If so I can do it here as well - otherwise it may make sense to treat this one differently.

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: The overhead in generated code should not be too large and the optimizer should reduce it to exactly the same code, but please test that. The main reason why I want to push inline assembly is because it makes the transition to webassembly easier.

We can discuss on wednesday whether we want 050 for that.

Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Not sure whether people are more likely to read Changelogs for 0.5.0 than for earlier releases - the problem with this change is that it changes semantics, but is absolutely silent. And I could actually imagine people using signed right shifts, while expecting to end up at zero at some point... An option would be to instead introduce a temporary warning about all signed right shifts before 0.5.0 (and maybe even for pragma 050 and only remove it with actual 0.5.0) - but a warning might be too strong for that... What do you think?

Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Ah OK - sorry, I missed your last comment.

**Comment by ekpyron** [comment]
@chriseth
Uses inline assembly now.
I had a look at the optimized EVM assembly output and at the gas estimates. The gas estimates are identical for the assembly and the non-assembly versions and the generated code is slightly different (as one would expect), but looks fine.

**Comment by ekpyron** [comment]
Note: still to be discussed: should this be a pragma 050 change or not?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 3]
```diff
@@ -1718,11 +1718,22 @@ void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator, Type co
...
-		if (m_context.evmVersion().hasBitwiseShifting() && !c_valueSigned)
-			m_context << Instruction::SHR;
+		if (m_context.evmVersion().hasBitwiseShifting())
+			m_context << Instruction::SAR;
 		else
-			m_context << u256(2) << Instruction::EXP << Instruction::SWAP1 << (c_valueSigned ? Instruction::SDIV : Instruction::DIV);
+		{
+			if (c_valueSigned)
+				m_context.appendInlineAssembly(R"({
+					let xor_mask := sub(0,slt(value_to_shift,0))
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Please add a space after each comma :)

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Can we assume the value to be clean, or do we have to call `signextend` before the check? Since we previously only used `SHR`, it should be fine, I think.

Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: I think it should be fine, but I can try to find some test cases for that and/or read through the code to see whether the assumption is valid. I think I'll also add a few more test cases for more odd and even, signed and unsigned values of different bit widths in general.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 4]
```diff
@@ -10133,9 +10150,29 @@ BOOST_AUTO_TEST_CASE(shift_right_negative_lvalue)
...
+{
+	char const* sourceCode = R"(
+			contract C {
+				function f(int16 a, int16 b) returns (int) {
+					return a >> b;
+				}
+			}
+		)";
+	compileAndRun(sourceCode, 0, "C");
+	ABI_CHECK(callContractFunction("f(int16,int16)", u256(-4266), u256(0)), encodeArgs(u256(-4266)));
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Would it be better to use hex here?

Comment (ekpyron) [reply test/libsolidity/SolidityEndToEndTest.cpp]: ``boost::multiprecision``'s doc states ``signed to unsigned conversions extract the last MaxBits bits of the 2's complement representation of the input value``, so negative decimals are fine. To me hex values with a negative sign look weird and using the 2's complement might be harder to understand (especially with rounding towards negative infinity).

The existing tests also already used a mixture of hex and dec values, so if we were to agree that hex is better, then we should probably change it everywhere.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 3 above)
Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Added some test cases that verify that ``signextend`` is not needed.

**Comment by axic** [comment]
@ekpyron do you want to pull out the test cases? Those can be merged now (with the current values).

**Comment by ekpyron** [comment]
Note: Needs to be rebased once the tests in #4102 are merged into develop and develop is merged into 050.

**Comment by ekpyron** [comment]
Rebased.

File: Changelog.md [hunk 5]
```diff
@@ -1,5 +1,8 @@
 ### 0.5.0 (unreleased)
 
+Breaking Changes:
+ * General: Signed right shift uses proper arithmetic shift, i.e. rounding towards negative infinity.
```
Comment (chriseth) [inline Changelog.md]: Can you please better highlight that this is a semantics change that might go unnoticed?

File: docs/types.rst [hunk 6]
```diff
@@ -60,15 +60,14 @@ operators are :ref:`literals<rational_literals>` (or literal expressions).
 Division by zero and modulus with zero throws a runtime exception.
 
 The result of a shift operation is the type of the left operand. The
-expression ``x << y`` is equivalent to ``x * 2**y``, and ``x >> y`` is
-equivalent to ``x / 2**y``. This means that shifting negative numbers
-sign extends. Shifting by a negative amount throws a runtime exception.
+expression ``x << y`` is equivalent to ``x * 2**y``, and, for positive integers,
+``x >> y`` is equivalent to ``x / 2**y``. For negative ``x``, ``x >> y``
+is equivalent to dividing by a power of ``2`` while rounding down (towards negative infinity).
```
Comment (chriseth) [inline docs/types.rst:65]: We should probably say something about literals. `-1>>2` will result in `-1/2` and thus `int(-1>>1)` will be different from `int(-1) >> 1`, right?

Comment (chriseth) [reply docs/types.rst:65]: There is also a comment in Types.cpp:1057 (BinaryOperatorResult of RationalNumberType).

Comment (ekpyron) [reply docs/types.rst:65]: I hadn't thought about literals actually - I think they need to be reworked as well.
Looking at Types.cpp:1060-1077 it looks like ``-1>>1`` will actually perform logical shift and not arithmetic shift and result in 0 right now. Similarly ``-1>>2``, etc.

The question is: *should* ``-1>>1`` be ``-1/2`` or should it be ``-1``?

Comment (chriseth) [reply docs/types.rst:65]: It should be consistent with `int(-1) >> 1` I would say. @axic what do you think?

Comment (ekpyron) [reply docs/types.rst:65]: I would agree that ``-1>>1`` should be consistent with ``int(-1) >> 1`` and therefore result in ``-1``. Fractional operands are disallowed anyways and one can always use ``-1/(2^1)`` (resp. ``-1/(2^c)`` instead of ``-1>>c``, where ``c`` is a positive literal) instead, if one wants fractional results.

In any case I will add test cases for right shifting negative literals - they were apparently missing or haven't been updated.

File: libsolidity/ast/Types.cpp [hunk 7]
```diff
@@ -1070,9 +1070,9 @@ TypePointer RationalNumberType::binaryOperatorResult(Token::Value _operator, Typ
 			{
 				uint32_t exponent = other.m_value.numerator().convert_to<uint32_t>();
 				if (exponent > mostSignificantBit(mp::abs(m_value.numerator())))
-					value = 0;
+					value = m_value.numerator() < 0 ? -1 : 0;
 				else
-					value = rational(m_value.numerator() / mp::pow(bigint(2), exponent), 1);
+					value = rational(m_value.numerator() >> exponent, 1);
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: It looks like boost can actually perform arithmetic right shift on ``bigint``, but I'm still looking for a mention of that in the boost docs and haven't found anything so far...

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Ok - I think this is undefined in boost, so I will have to change away from this.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 8]
```diff
@@ -1722,11 +1722,22 @@ void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator, Type co
...
-			m_context << Instruction::SHR;
+		if (m_context.evmVersion().hasBitwiseShifting())
+			m_context << Instruction::SAR;
 		else
-			m_context << u256(2) << Instruction::EXP << Instruction::SWAP1 << (c_valueSigned ? Instruction::SDIV : Instruction::DIV);
+		{
+			if (c_valueSigned)
+				m_context.appendInlineAssembly(R"({
+					let xor_mask := sub(0, slt(value_to_shift, 0))
+					value_to_shift := xor(div(xor(value_to_shift, xor_mask), exp(2, shift_amount)), xor_mask)
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Sorry, but I already forgot why this works. Can you perhaps add some comments?

File: libsolidity/ast/Types.cpp [hunk 9]
```diff
@@ -1073,9 +1073,14 @@ TypePointer RationalNumberType::binaryOperatorResult(Token::Value _operator, Typ
 			{
 				uint32_t exponent = other.m_value.numerator().convert_to<uint32_t>();
 				if (exponent > mostSignificantBit(mp::abs(m_value.numerator())))
-					value = 0;
+					value = m_value.numerator() < 0 ? -1 : 0;
 				else
-					value = rational(m_value.numerator() / mp::pow(bigint(2), exponent), 1);
+				{
+					if (m_value.numerator() < 0)
+						value = rational(-(mp::abs(m_value.numerator()) - 1) / mp::pow(bigint(2), exponent) - bigint(1), 1);
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Can you perhaps also add a comment explaining why this is the correct formula?

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Ok - I'll reformulate it slightly as well and hope it will be easier to see afterwards.

**Comment by ekpyron** [comment]
I think we actually hadn't talked about how I emulate SAR so far - are the comments I just added clear enough to explain what's happening?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 10]
```diff
@@ -10263,16 +10263,73 @@ BOOST_AUTO_TEST_CASE(shift_right_negative_lvalue)
...
+	ABI_CHECK(callContractFunction("f(int256,int256)", u256(-4267), u256(16)), encodeArgs(u256(-1)));
+	ABI_CHECK(callContractFunction("f(int256,int256)", u256(-4267), u256(17)), encodeArgs(u256(-1)));
+}
+
+BOOST_AUTO_TEST_CASE(shift_right_negative_literal)
+{
+	char const* sourceCode = R"(
+			contract C {
+				function f1() pure returns (int) {
+					return -4266 >> 0;
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Can you change these tests to move the expectations closer to the test case, perhaps using `return (-4266 >> 0 == -4266)`?

**Comment by chriseth** [comment]
@axic @leonardoalt can you please review the shifting routines?

**Review by axic — DISMISSED** [review]
This in general looks good to me.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 11]
```diff
@@ -1737,11 +1737,28 @@ void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator, Type co
...
+			if (c_valueSigned)
+				// For negative values xor_mask has all bits set and xor(value_to_shift, xor_mask) will be
+				// the bitwise complement of value_to_shift, i.e. abs(value_to_shift) - 1. Dividing this by
+				// exp(2, shift_amount) results in a value that is positive and strictly smaller than the
+				// absolute value of the desired result. Taking the complement again changes the sign
+				// back to negative and subtracts one, resulting in rounding towards negative infinity.
+				// For positive values xor_mask is zero and xor(value_to_shift, xor_mask) is again value_to_shift.
+				m_context.appendInlineAssembly(R"({
+					let xor_mask := sub(0, slt(value_to_shift, 0))
+					value_to_shift := xor(div(xor(value_to_shift, xor_mask), exp(2, shift_amount)), xor_mask)
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: By the way, something that might also work is (we could check if it is cheaper):

```
mstore(0x20, value)
signextend(mload(sub(0x20, shift_amount)), shift_amount)
```

Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: ``signextend`` uses byte offsets, not bit offsets, doesn't it?

Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Not sure how I could get a shift that's not a multiple of 8 this way...

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Oh you are right! I forgot about that!

**Comment by ekpyron** [comment]
@chriseth Do you think the comment in ExpressionCompiler.cpp works now?

**Comment by axic** [comment]
I am still fine merging this. The code can be improved/changed, if needed, at a later time.

**Review by axic — DISMISSED** [review]

**Comment by axic** [comment]
This is in conflict now. Will rebase.

**Comment by axic** [comment]
Rebased. @chriseth are we ok to merge this?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 12]
```diff
@@ -1737,11 +1737,36 @@ void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator, Type co
 			m_context << u256(2) << Instruction::EXP << Instruction::MUL;
 		break;
 	case Token::SAR:
-		// NOTE: SAR rounds differently than SDIV
-		if (m_context.evmVersion().hasBitwiseShifting() && !c_valueSigned)
-			m_context << Instruction::SHR;
+		if (m_context.evmVersion().hasBitwiseShifting())
+			m_context << Instruction::SAR;
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Hm, I think this is broken. This should check for `c_valueSigned`, otherwise right shifting an unsigned value with the top bit set will be wrong.

**Review by axic — DISMISSED** [review]

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 12 above)
Comment (ekpyron) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Oh, wow - you're right! I was so focused on the emulation part that I missed this entirely! I fixed it and added another test case that would have caught this mistake.

**Review by axic — APPROVED** [review]
